Generate a initializer file when installing the gem

thisiscab opened the pull request:
This would resolve https://github.com/glossier/solidus_retail/issues/39

I felt like that there was a need to create a configuration object that would be configured in the host app from an initializer.

thisiscab committed:
Add the initialize generator

This also removes the old namespace that we are not using anymore.

thisiscab committed:
Create a Spree Retail configuration object

This will hold all the related configuration for this gem to work.
That object will be fed from the value added by the initializer.

thisiscab committed:
Create the initializer when installing the gem

Instead of having to run an extra-command, this file is REQUIRED for the
gem to work, so I've decided to add it directly in the install
generator.

thisiscab committed:
Add the proper namespace convention for the generators

thisiscab committed:
Add the template path in the source path

This is required in order for the gem to find the initializer's template

thisiscab commented:
@braidn @DanielWright @bryanmtl 
:D Thank you!

thisiscab commented:
I am unsure how to deal with default values, so that the specs can at least run on locally and on Travis.

Since Shopify doesn't allow a "test" shop we have to use our own credential.

DanielWright commented:
As part of the test setup, can you generate an initializer?

thisiscab commented:
Then we would have to explicitly ask the gem user to enter its credential somewhere, there is an extra layer of overhead that we need to have, I am just unsure of the best way to put it.

DanielWright commented:
Imagine yourself integrating this extension into an app (and try to forget you've built the extension yourself). What would be the most compassionate way to capture their credentials? I think a good first step would be to generate an initializer that demonstrates how one might configure the extension. Think how Devise creates an initializer with sensible defaults. Does that make sense?

thisiscab commented:
Have you looked at the PR, isn't it what we are already doing? 

I see your point that for the gem itself, we should generate an initializer which I agree. My counter-argument is that there are no "sensible defaults" the user of this extension will still have to enter their Shopify API key and Token in order to run the specs.

DanielWright commented:
Ah! I misunderstood your question, which is a shame cos it's a really really good one. Hmm.

I don't really like the way [`solidus_gateway`](https://github.com/solidusio/solidus_gateway/blob/master/spec/features/stripe_checkout_spec.rb#L14-L15) provides its own credentials, because I've actually run into a bug because of it (the API version those credentials run is out-of-date, which confounded me for _days_).

I think we have to pull them from `ENV` for the specs. We can't guarantee any credentials we hard-code will remain functional. So, what do you think about depending on ENV vars, but with a human-friendly error-message if they're not found?

Something like:

```
begin
  ENV.fetch('SHOPIFY_API_KEY')
rescue KeyError
  puts "You must set the following environment variables: ... Please see <a href>the documentation</a> for ways to accomplish this."
end
```

thisiscab committed:
Throw error if there is a missing env variable

We need for the user of the gem to set-up their own credential in order
to run the specs. If there are no credentials set in the environment
variables then we can't run the specs and we output an error message.

thisiscab committed:
Document that feature in the readme

thisiscab commented:
What do you think of the changes?
@DanielWright

thisiscab committed:
Re-add webmock/rspec spec require

This is required in order to use `stub_request`.

braidn commented on lib/generators/spree/retail/install/templates/config/initializers/spree_retail.rb:
```diff
@@ -0,0 +1,4 @@
+# Spree Retail Setup
```
Just a quick naming question here. Since this exact gem will be called 'solidus_retail' should the initializer have the same name?

thisiscab replied:
@DanielWright What do you think of this? 
Everywhere we use `spree` instead of `solidus`.

braidn commented on spec/support/simplecov.rb:
```diff
@@ -0,0 +1,21 @@
+# Run Coverage report
```
Is in the scope of this PR?

braidn commented on README.md:
````diff
@@ -38,8 +38,20 @@ Simply add this require statement to your spec_helper:
+SHOPIFY_SHOP_NAME=
+```
 
+Copyright (c) 2016 [name of extension creator], released under the New BSD License
````
Should we set this as Glossier?

braidn reviewed (commented):
CRed! ‼️ 🐩

braidn commented:
Also @cabouffard you never pinged me here for a CR. Did you read my [post from yesterday morning](https://dynamomtl.slack.com/files/braden/F2H0XMYE4/Keys_To_A_Successful_Code_Review_Request)?

braidn commented:
Also, it would be super nice to move #39 due to it's conversation into here when creating the PR, unless you don't think that transition makes sense. Helps keep context in one thread IMHO.

thisiscab committed:
Change creator name to Glossier in the README

thisiscab committed:
Merge branch 'master' into feature/add-initializer-generator
